olive/data/component/pre_process_data.py

```diff
@@ -122,7 +122,119 @@ def _tokenizer_and_align_labels(examples):
+    tokenized_datasets = _huggingface_pre_process_helper(
         _dataset, model_name, input_cols, label_cols, _tokenizer_and_align_labels, **kwargs
     )
     return BaseDataset(tokenized_datasets, label_cols=["label"], max_samples=max_samples)
+
+
+@Registry.register_pre_process()
+def text_generation_huggingface_pre_process(
+    _dataset, model_name, input_cols, seqlen, stride=None, max_samples=None, **kwargs
+):
+    """Pre-process data.
+
+    Args:
+        _dataset (object): Data to be pre-processed.
+        model_name (str): Name of the huggingface model.
+        input_cols (list): List of input columns.
+        seqlen (int): Length of the sequence.
+        stride (int): Stride to use when splitting the sequence. I
+            If None, the sequence is split into non-overlapping sequences. No context is used.
+            When stride is not None, we use a sliding window approach to split the sequence. The stride is
+            also used as context length.
+        max_samples (int): Maximum number of samples to use.
+        **kwargs: Additional arguments.
+            joiner (str): Delimiter to use when joining the rows of the input columns.
+            random_seed (int): Random seed to use. If not None, we use the random seed to choose the starting
+                point of the sequence.
+
+    Returns:
+        object: Pre-processed data.
+    """
+    from random import Random
+
+    from datasets import Dataset as HFDataset
+    from transformers import AutoTokenizer
+
+    # get tokenizer
+    tokenizer = AutoTokenizer.from_pretrained(model_name)
+
+    # gather text from all input columns
+    text_list = []
+    for input_col in input_cols:
+        text_list += _dataset[input_col]
+    # delimiter between the text sequences
+    joiner = kwargs.get("joiner", " ")
+    text = joiner.join(text_list)
+
+    # in order to make processing faster we will only tokenize as much as needed
+    # assumes that num words > num tokens
+    split_text = text.split(" ")
+    num_text = len(split_text)
+
+    tokenized_inputs = {
+        "input_ids": [],
+        "target_ids": [],
+    }
+
+    random_seed = kwargs.get("random_seed")
+    if random_seed is None:
+        # no randomization, just use contiguous blocks of tokens
+        max_text = max_samples * seqlen if max_samples is not None else num_text
+        max_text = min(max_text, num_text)
+        encodings = tokenizer(" ".join(split_text[:max_text]), return_tensors="pt")
+        input_ids = encodings.input_ids
+
+        num_tokens = input_ids.shape[1]
+        step = stride or seqlen
+        # loop over the number of tokens
+        # all inputs must be seqlen long
+        for begin_loc in range(0, num_tokens - seqlen, step):
+            # end_loc is the beginning of the next sequence
+            end_loc = begin_loc + seqlen
+            # get the input sequence
+            input_ids = encodings.input_ids[0, begin_loc:end_loc].clone()
```

Comment:
Why we only get the first sentence here? `input_ids[0, a:b]` should be `input_ids[:, a:b]`?

Reply:
`encodings.input_ids` is has the batch dimension but it is always `1` in this case since we only give one `text`. We don't use `dataset.map` here that takes care of the batch dimension so we are removing the batch dimension. it gets added back by the dataloader based on the batchsize later.

Reply:
Thanks, I was thinking that you handle the sentence one-by-one. But as you mentioned above, there is only one text input for tokenizer.